src/security/core/src/dds_security_serialize.c

```diff
@@ -0,0 +1,864 @@
+DDS_Security_Serialize_OctetSeq(
+     DDS_Security_Serializer ser,
+     const DDS_Security_OctetSeq *seq)
+{
+     DDS_Security_Serialize_uint32_t(ser, seq->_length);
+     serbuffer_adjust_size(ser, seq->_length);
+     memcpy(&(ser->buffer[ser->offset]), seq->_buffer, seq->_length);
+     ser->offset += seq->_length;
+}
+
+static void
+DDS_Security_Serialize_BinaryProperty(
+     DDS_Security_Serializer ser,
+     const DDS_Security_BinaryProperty_t *property)
+{
+    DDS_Security_Serialize_string(ser, property->name);
+    DDS_Security_Serialize_OctetSeq(ser, &property->value);
+}
+
+void
+DDS_Security_Serialize_PropertySeq(
+     DDS_Security_Serializer ser,
+     const DDS_Security_PropertySeq *seq)
+{
+    uint32_t i;
+
+    DDS_Security_Serialize_uint32_t(ser, seq->_length);
+    for (i = 0; i < seq->_length; i++) {
+         DDS_Security_Serialize_Property(ser, &seq->_buffer[i]);
+    }
+}
+
+void
+DDS_Security_Serialize_BinaryPropertyArray(
+     DDS_Security_Serializer serializer,
+     const DDS_Security_BinaryProperty_t **properties,
+     const uint32_t propertyLength)
+{
+    uint32_t i;
+
+     DDS_Security_Serialize_uint32_t(serializer, propertyLength);
+     for (i = 0; i < propertyLength ; i++) {
+          DDS_Security_Serialize_BinaryProperty(serializer, properties[i]);
+     }
+}
+
+void
+DDS_Security_Serialize_BinaryPropertySeq(
+     DDS_Security_Serializer serializer,
+     const DDS_Security_BinaryPropertySeq *seq)
+{
+    uint32_t i;
+
+    DDS_Security_Serialize_uint32_t(serializer, seq->_length);
+    for (i = 0; i < seq->_length; i++) {
+         DDS_Security_Serialize_BinaryProperty(serializer, &seq->_buffer[i]);
+    }
+}
+
+
+static void
+DDS_Security_Serialize_DataHolder(
+     DDS_Security_Serializer ser,
+     const DDS_Security_DataHolder *holder)
+{
+    DDS_Security_Serialize_string(ser, holder->class_id);
+    DDS_Security_Serialize_PropertySeq(ser, &holder->properties);
+    DDS_Security_Serialize_BinaryPropertySeq(ser, &holder->binary_properties);
+}
+
+
+static void
+DDS_Security_Serialize_BuiltinTopicKey(
+     DDS_Security_Serializer ser,
+     DDS_Security_BuiltinTopicKey_t key)
+{
+    serbuffer_align(ser, sizeof(uint32_t));
+    DDS_Security_Serialize_uint16(ser, PID_PARTICIPANT_GUID);
+    DDS_Security_Serialize_uint16(ser, 16);
+    DDS_Security_Serialize_uint32_t(ser, key[0]);
+    DDS_Security_Serialize_uint32_t(ser, key[1]);
+    DDS_Security_Serialize_uint32_t(ser, key[2]);
+    /* 4 Bytes are expected for whatever reason (gid vs guid?). */
+    DDS_Security_Serialize_uint32_t(ser, 0);
+}
+
+static void
+DDS_Security_Serialize_UserDataQosPolicy(
+     DDS_Security_Serializer ser,
+     DDS_Security_OctetSeq *seq)
+{
+    if (seq->_length > 0) {
+        serbuffer_align(ser, sizeof(uint32_t));
+        DDS_Security_Serialize_uint16(ser, PID_USER_DATA);
+        DDS_Security_Serialize_uint16(ser, (unsigned short)seq->_length);
+        DDS_Security_Serialize_OctetSeq(ser, seq);
+    }
+}
```

Comment:
DDS_Security_Serialize_UserDataQosPolicy() seems to be based on an older version.

Reply:
Where can I find the new one?